### PR TITLE
Add mixpanel tracking to transporter promos

### PR DESCRIPTION
**Pull request by davidpmccormick**
## What is this PR trying to achieve?
Adding mixpanel tracking to transporter promos.

screen shot 2017-05-17 at 15 12 33

### CODE REVIEW

File: client/js/components/async-content.js [hunk 1]
```diff
@@ -37,6 +39,15 @@ export default function asyncContent(el, dispatch) {
           containImages: false,
           scrollToClickedItem: false
         });
+
+        const transporterPromos = document.querySelectorAll('[data-track="transporter-promo"]');
+        nodeList(transporterPromos).forEach((transporterPromo) => {
+          const url = transporterPromo.href;
+
+          transporterPromo.addEventListener('click', () => {
+            trackEvent({name: 'Transporter promo', properties: {url}});
```
Comment (jamesgorrie) [inline client/js/components/async-content.js]: As the content in this is eventually going to be dynamic, I'm not sure the URL is massively useful.
Potentially position is though?

Comment (davidpmccormick) [inline client/js/components/async-content.js]: Good point – would you take the URL out (i.e. is there any negative to it staying in? Drowning in data, maybe?).

File: client/js/components/async-content.js [hunk 2]
```diff
@@ -37,6 +39,15 @@ export default function asyncContent(el, dispatch) {
           containImages: false,
           scrollToClickedItem: false
         });
+
+        const transporterPromos = document.querySelectorAll('[data-track="transporter-promo"]');
```
Comment (jamesgorrie) [inline client/js/components/async-content.js]: Are we not tracking the Transporter in this case?

**Comment by jamesgorrie** [comment]
I wonder if we could use this as an opportunity to make something more generic for the rest of the components, as component tracking is something I imagine we want in the future.

Also I am not sure what we are tracking as a component, is it the Transporter (a list of promos), or the individual promos?

Being pedantic as it's hard to undo tracking metrics.

File: client/js/components/async-content.js (hunk 2 above)
Comment (davidpmccormick) [inline client/js/components/async-content.js]: Yeah, I think you're right. Will update.

File: client/js/util.js [hunk 3]
```diff
@@ -41,3 +41,31 @@ export {
...
+
+    possibleTargets.forEach((possibleTarget) => {
+      const correctTarget = checkTarget(delegateEl, eventEl, possibleTarget);
+
+      if (correctTarget) {
+        return fn.call(correctTarget, event);
+      }
+    });
+  });
+}
```
Comment (jamesgorrie) [inline client/js/util.js]: As this does a DOM read every time - do you think this might be a good place to add fastDOM?

File: client/js/tracking.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  const promos = nodeList(transporter.querySelectorAll('.promo'));
+  const promo = event.target.closest('.promo');
+  const link = promo ? promo.href : null;
+  const position = promo ? promos.indexOf(promo) : null;
+  const properties = {id: transporter.id, link, position};
+
+  if (link) {
+    trackEvent({name: 'Transporter', properties});
+  }
+});
```
Comment (jamesgorrie) [inline client/js/tracking.js]: Thoughts for potentially another PR.

I wonder if there is something more generic we can do here, potentially adding this into the DOM:
```twig
<div data-component="Transformer" data-component-id="12">
  {% for item in items %}
  <a data-track-click='{ "position": loop.index }'>Things</a>
  {% .endfor %}
</div>
```

Then when someone clicks that - we can figure out the component and it's ID (something that will probably be generic)* and send over the data.

As mentioned this wields the same results, so no matter until we do it again.

* I wonder if this is a chance to fix my mistake and call it `componentId`? We can loose a bit of data around the image galleries.

File: client/js/app.js [hunk 5]
```diff
@@ -17,6 +17,7 @@ import asynContent from './components/async-content';
 import contentSlider from './components/content-slider';
 import gaScrollDepth from '../libs/ga-scroll-depth';
 import joinCohort from './components/join-cohort';
+import './tracking';
```
Comment (jamesgorrie) [inline client/js/app.js]: Could we `init` this for consistency (and it's always a little unclear that if you include something, it's run)

File: client/js/tracking.js (hunk 4 above)
Comment (davidpmccormick) [inline client/js/tracking.js]: Really like this idea, yep.

Comment (jamesgorrie) [inline client/js/tracking.js]: Are you going to give something like this a go?
Just wondering if I wait on that before reviewing - if so, maybe mark the PR as WIP.

Comment (davidpmccormick) [inline client/js/tracking.js]: Was thinking of doing it in another PR, but perhaps doing it here is the better idea.

File: client/js/util.js (hunk 3 above)
Comment (davidpmccormick) [inline client/js/util.js]: I think that adding fastDOM is a chunk of work that probably belongs in its own PR.

**Comment by jamesgorrie** [comment]
@davidpmccormick are we good to review this?
If so I'll take a look once the conclicts are fixed.

**Comment by davidpmccormick** [comment]
@jamesgorrie yes, re-review please.

File: server/views/components/numbered-list/index.njk [hunk 6]
```diff
@@ -12,7 +12,8 @@
     </header>
   {% endif %}
   <ol class="numbered-list__list{% if data.sliderId %} js-numbered-slider{% endif %} {{ 'grid grid--dividers' if isTransporter }}"
-    {% if data.sliderId %} id="{{data.sliderId}}" {% endif %}>
+    {% if isTransporter %}data-component="Transporter" {% endif %}
```
Comment (jamesgorrie) [inline server/views/components/numbered-list/index.njk]: Could we not make this more generic, something like `componentName`?

File: server/views/components/numbered-list/index.njk [hunk 7]
```diff
@@ -24,7 +25,9 @@
           <span class="numbered-list__number" aria-hidden="true">{{ loop.index }}</span>
 
           {% if isTransporter %}
-            {% componentV2 'promo', item %}
+            <span data-track-click='{"position": {{ loop.index }}, "url": "{{ item.url }}" }'>
```
Comment (jamesgorrie) [inline server/views/components/numbered-list/index.njk]: Neato.
